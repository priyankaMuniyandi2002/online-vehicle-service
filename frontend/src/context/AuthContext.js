/** TAC SERVICE BOOKING APP - REACT AUTH CONTEXT FILE **/
/*
 * React Context API used to manage global state for user authentication and authorization in the TAC Service Booking app.
 */

import { createContext, useReducer, useEffect } from "react";

export const AuthContext = createContext();

export const authReducer = (state, action) => {
  switch (action.type) {
    case "LOGIN":
      return { user: action.payload };
    case "LOGOUT":
      return { user: null };
    default:
      return state;
  }
};

export const AuthContextProvider = ({ children }) => {
  const [state, dispatch] = useReducer(authReducer, {
    user: null,
  });

  useEffect(() => {
    const user = JSON.parse(localStorage.getItem("user"));

    if (user) {
      dispatch({ type: "LOGIN", payload: user });
    }
  }, []);

  /* Check if running in a test environment */
  const isTestEnvRunning = process.env.NODE_TEST_ENV === "test";

  /* Log the auth context state only if not in a test environment */
  !isTestEnvRunning && console.log("AuthContext state:", state);

  return (
    <AuthContext.Provider value={{ ...state, dispatch }}>
      {children}
    </AuthContext.Provider>
  );
};
