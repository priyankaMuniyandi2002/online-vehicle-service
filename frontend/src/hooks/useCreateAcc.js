/** TAC SERVICE BOOKING APP - CUSTOM REACT HOOK FOR CREATING USER ACCOUNT **/

import { useNavigate } from "react-router-dom";
import { useAuthContext } from "./useAuthContext";

export const useCreateAcc = () => {
  const navigate = useNavigate();
  const { dispatch } = useAuthContext();

  const createAcc = async (createAccCredentials) => {
    const { rememberMe, ...createAccCredentialsToSend } = createAccCredentials;

    try {
      const response = await fetch("http://localhost:8080/api/create-acc", {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(createAccCredentialsToSend),
      });

      const json = await response.json();
      console.log(json);

      if (response.ok) {
        navigate("/");
        if (rememberMe) {
          // If "rememberMe" is checked - save the relevant user details to local storage when the user creates account successfully.
          localStorage.setItem("user", JSON.stringify(json));
        } else {
          // If "Remember Me" is not checked, remove the preference from localStorage
          localStorage.removeItem("user");
        }

        // update the auth context global state.
        dispatch({ type: "LOGIN", payload: json });
      }
    } catch (error) {
      console.log("Error", error);
    }
  };

  return { createAcc };
};
