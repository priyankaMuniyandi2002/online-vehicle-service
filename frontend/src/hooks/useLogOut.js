/** TAC SERVICE BOOKING APP - CUSTOM REACT HOOK FOR USER LOGOUT **/

import { useAuthContext } from "./useAuthContext";

export const useLogOut = () => {
  const { dispatch } = useAuthContext();

  const logOut = () => {
    // removing the user item from local storage when user is logged out.
    localStorage.removeItem("user");

    // dispatch logOut action - authContext.js.
    dispatch({ type: "LOGOUT" });
  };

  return { logOut };
};
