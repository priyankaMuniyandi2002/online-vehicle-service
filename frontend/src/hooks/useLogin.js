/** TAC SERVICE BOOKING APP - CUSTOM REACT HOOK FOR USER LOGIN **/

import { useState } from "react";
import { useAuthContext } from "./useAuthContext";
import { useNavigate } from "react-router-dom";

export const useLogin = () => {
  const navigate = useNavigate();
  const [loginError, setLoginError] = useState(null);
  const { dispatch } = useAuthContext();

  const login = async (loginCredentials) => {
    const { rememberMe, ...loginCredentialsToSend } = loginCredentials;
    setLoginError(null);

    try {
      const response = await fetch("http://localhost:8080/api/login", {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(loginCredentialsToSend),
      });

      const json = await response.json();

      

      if (!response.ok) {
        console.log(json.error);
        setLoginError(json.error);
      }

      if (response.ok) {
        navigate("/");

        if (rememberMe) {
          // If "rememberMe" is checked - save the relevant user details to local storage when the user is logged into to their account.
          localStorage.setItem("user", JSON.stringify(json));
        } else {
          // If "Remember Me" is not checked, remove the preference from localStorage
          localStorage.setItem("user", JSON.stringify(json));

        }

        // update the auth context global state.
        dispatch({ type: "LOGIN", payload: json });
      }
    } catch (error) {
      console.log("Error", error);
    }
  };

  return { login, loginError };
};
