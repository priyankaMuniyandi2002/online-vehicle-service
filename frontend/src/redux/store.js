/** TAC SERVICE BOOKING APP REDUX TOOLKIT STORE **/
/* This Redux Toolkit store handles the global state in the TAC Service Booking App. */

import { configureStore } from "@reduxjs/toolkit";
import bookingsReducer from "./bookingsSlice";
import helpModalReducer from "./helpModalSlice";
import deleteModalReducer from "./deleteConfirmModalSlice";

/* Configuring the Redux toolkit store with multiple reducers to be used globally in the app. */
const store = configureStore({
  reducer: {
    bookings: bookingsReducer,
    helpModal: helpModalReducer,
    deleteModal: deleteModalReducer,
  },
});

export default store;
