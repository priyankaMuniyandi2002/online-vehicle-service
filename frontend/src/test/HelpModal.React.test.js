/** TAC SERVICE BOOKING APP HELP MODAL JEST SNAPSHOT TEST FILE **/

import React from "react";
import renderer from "react-test-renderer";
import { AuthContextProvider } from "../context/AuthContext";
import { useAuthContext } from "../hooks/useAuthContext";
import { Provider } from "react-redux";
import store from "../redux/store";
import HelpModal from "../components/modal-components/HelpModal";

/* Mock the useAuthContext hook */
jest.mock("../hooks/useAuthContext");

describe("Ensuring the Proper Rendering of the HelpModal Component", () => {
  it("displays as expected for authenticated users", () => {
    /* Mock authenticated user - service booking agent */
    const mockUser = {
      firstName: "Jay",
      lastName: "Shetty",
      email: "jay.shetty@.tac.co.za",
      token:
        "pqaJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJsbFQiOiJKb2huIiwibGFzdE5hbWUiOiJEb2UiLCJlqSdapbCI6ImpvaG5kb2VAZXhhbXBsZS5jb20ifQ.1234567890abcdef",
    };

    /* Mock the behavior of custom useAuthContext hook - returning the relevant data of an authenticated user */
    useAuthContext.mockReturnValue({ user: mockUser });

    /* Render the component within the AuthContextProvider and Redux Providers */
    const wrapper = (
      <AuthContextProvider>
        <Provider store={store}>
          <HelpModal />
        </Provider>
      </AuthContextProvider>
    );

    /* Create a snapshot */
    const tree = renderer.create(wrapper).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
