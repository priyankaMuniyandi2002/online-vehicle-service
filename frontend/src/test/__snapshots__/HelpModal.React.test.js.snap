// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ensuring the Proper Rendering of the HelpModal Component displays as expected for authenticated users 1`] = `
<section
  className="modal-section-hidden"
>
  <div
    className="help-modal-container"
  >
    <div
      className="title-container"
    >
      <h2
        className="title"
      >
        Help Menu
      </h2>
    </div>
    <span
      className="modal-close-btn"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium icon close-icon css-i4bv87-MuiSvgIcon-root"
        data-testid="CloseIcon"
        focusable="false"
        onClick={[Function]}
        viewBox="0 0 24 24"
      >
        <path
          d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </span>
    <p>
      This app allows the service booking agent to schedule and track service bookings.
    </p>
    <div
      className="title-container"
    >
      <h4
        className="help-modal-title"
      >
        How to Schedule a Service Booking
      </h4>
    </div>
    <p>
      To schedule a new service booking in your app, follow these step-by-step guidelines:
    </p>
    <ol>
      <li>
        <strong>
          Initiate a New Booking:
        </strong>
         While on your Booking Dashboard, locate and click on the "Schedule Booking" option. This action will redirect you to the Schedule Booking Page.
      </li>
      <li>
        <strong>
          Provide Customer Details:
        </strong>
         On the Schedule Booking Page, you'll need to fill in essential customer details. These include the customer name, email address and contact number. Additionally, you'll need to fill in the customer's vehicle make, model and registration number.
      </li>
      <li>
        <strong>
          Select the Service Option:
        </strong>
         In the drop-down menu, select the specific service that needs to be delivered to the customer.
      </li>
      <li>
        <strong>
          Set Service Date and Time:
        </strong>
         Specify the date and time for the service appointment, that is the most convenient time for the customer.
      </li>
      <li>
        <strong>
          Additional Information:
        </strong>
         If there are any additional notes or special requirements related to the booking, include them in the designated section.
      </li>
      <li>
        <strong>
          Confirm Booking:
        </strong>
         After filling out all the required information, review your booking details to ensure accuracy. When you're satisfied with the information provided, click on the "CONFIRM BOOKING" button to confirm the booking.
      </li>
      <li>
        <strong>
          View Booking Confirmation:
        </strong>
         Once the booking is successfully scheduled, you will receive a confirmation message, and the new service booking will be reflected on your Booking Dashboard.
      </li>
    </ol>
    <div
      className="title-container"
    >
      <h4
        className="help-modal-title"
      >
        How to Update a Service Booking
      </h4>
    </div>
    <p>
      To update an existing service booking, follow these step-by-step guidelines:
    </p>
    <ol>
      <li>
        <strong>
          Locate the Booking to Update:
        </strong>
         Scroll through your Booking Dashboard to find the specific booking you wish to update.
      </li>
      <li>
        <strong>
          Initiate Booking Update:
        </strong>
         In the booking display, navigate to the action items column for the booking you want to update. Click on the "Update" icon associated with the booking you want to update. This action will redirect you to the Update Booking Page.
      </li>
      <li>
        <strong>
          Review Pre-Populated Details:
        </strong>
         On the Update Booking Page, you'll find the existing booking details pre-populated for your reference.
      </li>
      <li>
        <strong>
          Update Customer Information:
        </strong>
         If needed, update the customer's name, email address, and contact number. Additionally, modify the customer's vehicle make, model, and registration number, as required
      </li>
      <li>
        <strong>
          Update Service Details:
        </strong>
         If the service type needs to be changed, use the provided drop-down menu to select the appropriate service.
      </li>
      <li>
        <strong>
          Update Service Date and Time:
        </strong>
         Modify any changes to the service appointment date and time, ensuring it suits the customer's convenience.
      </li>
      <li>
        <strong>
          Additional Information:
        </strong>
         If there are any additional notes or special requirements related to the booking, include them in the designated section.
      </li>
      <li>
        <strong>
          Confirm Booking Update:
        </strong>
         After making the necessary updates, review the information to ensure accuracy. When you're satisfied with the information provided, click on the "UPDATE BOOKING" button to complete the booking update.
      </li>
      <li>
        <strong>
          View Updated Booking Confirmation:
        </strong>
         Once the booking is successfully updated, you will receive a confirmation message, and the updated service booking will be reflected on your Booking Dashboard.
      </li>
    </ol>
    <div
      className="title-container"
    >
      <h4
        className="help-modal-title"
      >
        How to Cancel a Service Booking
      </h4>
    </div>
    <p>
      To cancel a service booking, follow these step-by-step guidelines:
    </p>
    <ol>
      <li>
        <strong>
          Locate the Booking to Cancel:
        </strong>
         Scroll through your Booking Dashboard to find the specific booking you wish to cancel.
      </li>
      <li>
        <strong>
          Initiate Booking Cancellation:
        </strong>
         In the booking display, navigate to the action items column for the booking you want to cancel. Click on the "Cancel" icon associated with the booking you want to cancel.
      </li>
      <li>
        <strong>
          Confirmation Popup:
        </strong>
         Upon clicking the "Cancel" icon, a confirmation popup will appear with options:
        <ul>
          <li>
            <strong>
              Proceed with Cancellation:
            </strong>
             Click on "Proceed with Cancellation" to process the cancellation.
          </li>
          <li>
            <strong>
              Return to Dashboard:
            </strong>
             Click on "Return to Dashboard" to cancel the cancellation process and return to the Booking Dashboard.
          </li>
        </ul>
      </li>
      <li>
        <strong>
          View Booking Dashboard:
        </strong>
         Once the booking is successfully cancelled, you will receive a confirmation message, and the service booking will be removed from your Booking Dashboard.
      </li>
    </ol>
    <div
      className="title-container"
    >
      <h4
        className="help-modal-title"
      >
        How to Mark a Service Booking as Complete
      </h4>
    </div>
    <p>
      To mark a service booking as complete, follow these step-by-step guidelines:
    </p>
    <ol>
      <li>
        <strong>
          Locate the Booking to Mark as Complete:
        </strong>
         Scroll through your Booking Dashboard to find the specific booking you wish to mark as complete.
      </li>
      <li>
        <strong>
          Initiate Booking Completion:
        </strong>
         In the booking display, navigate to the action items column for the booking you want to mark as complete. Click on the "Complete" icon associated with the booking you want to mark as complete.
      </li>
      <li>
        <strong>
          View Booking Dashboard:
        </strong>
         Once the booking is successfully marked as completed, you will receive a confirmation message, and the service booking will be removed from your Booking Dashboard.
      </li>
    </ol>
    <div
      className="title-container"
    >
      <h4
        className="help-modal-title"
      >
        How to Search for a Service Booking
      </h4>
    </div>
    <p>
      To quickly find a specific service booking, follow these guidelines:
    </p>
    <ol>
      <li>
        <strong>
          Locate the Search Bar:
        </strong>
         On the Booking Dashboard, locate the search bar at the top of the page.
      </li>
      <li>
        <strong>
          Enter Search Criteria:
        </strong>
         Type in relevant information such as Work Order No. into the search bar. Alternatively, you can paste the search query if you copied it from elsewhere.
      </li>
      <li>
        <strong>
          View Search Results:
        </strong>
         As you type or paste, the app will dynamically display matching results in real-time.
      </li>
    </ol>
  </div>
</section>
`;
